docs: add docs and changelog for new terraform test state key attribute (#36322)
hashicorp/terraform: website/docs/language/tests/index.mdx

| `expect_failures`         | An optional attribute.                                                                                                   |               |
| `state_key`               | An optional attribute.                                                                                                   |               |

The `command` attribute and `plan_options` block tell Terraform which command and options to execute for each run block. The default operation, if you do not specify a `command` attribute or the `plan_options` block, is a normal Terraform apply operation.

The `command` attribute states whether the operation should be a [`plan`](/terraform/cli/commands/plan) or an [`apply`](/terraform/cli/commands/apply) operation.

The `plan_options` block allows test authors to customize the [planning mode](/terraform/cli/commands/plan#planning-modes) and [options](/terraform/cli/commands/plan#planning-options) they would typically need to edit via command-line flags and options. We cover the `-var` and `-var-file` options in the [Variables](#variables) section.

> The `state_key` allows for fine-grained control over which internal state file should be used for a given run block. This is discussed in more detail within [Modules State](#modules-state).

Reviewer: ```suggestion
The `state_key` allows for fine-grained control over which internal state file Terraform uses for a given run block. Refer to [Modules State](#modules-state) for more information.
```

### Modules state

> While Terraform executes a `terraform test` command, Terraform maintains at least one, but possibly many, state files within memory for each test file. Each internal state file is assigned a state key that Terraform uses internally to track the state file. The state key is a unique identifier for the state file and can be overridden by the `state_key` attribute of a `run` block.

Reviewer: ```suggestion
While Terraform executes a `terraform test` command, Terraform maintains at least one, but possibly many, state files within memory for each test file. Terraform assigns each internal state file a state key that it uses internally to track the state file. The state key is a unique identifier for the state file and you can override it with the `state_key` attribute of a `run` block.
```

There is always at least one state file that maintains the state of the main configuration under test. This state file is shared by all `run` blocks that do not have a `module` block specifying an alternate module to load. By default, there is also one state file per alternate module that Terraform loads. An alternate module state file is shared by all `run` blocks that execute the given module.

> The `state_key` attribute can override this default behavior, and force Terraform to use a specific state file for a given `run` block. This is useful when you want to share state between `run` blocks that do not reference the same module. The default state key is the `module` block's `source` attribute, or the empty string (`""`) if no alternate module is specified.

Reviewer: ```suggestion
You can override this default behavior with the `state_key` attribute and force Terraform to use a specific state file for a given `run` block. This is useful when you want to share state between `run` blocks that do not reference the same module. The default state key is the `module` block's `source` attribute, or the empty string (`""`) if no alternate module is specified.
```
